# Deep Learning Assignment

*Measuring Prediction Errors using Loss Functions: MSE and Cross-Entropy*

## 1. Objective

To understand how neural networks measure prediction errors by implementing two basic loss functions, Mean Squared Error (MSE) and Cross-Entropy Loss, and by comparing the difference between actual and predicted values on a simple classification problem.

## 2. Introduction

A loss function is a mathematical formula that converts the difference between the actual value and the value predicted by a model into a single number. A small loss means the predictions are close to the actual labels, and a large loss means they are far away. During training, a neural network updates its weights to minimise this number, so the loss function guides the whole learning process.

## 3. Task 1: Data (Actual Labels and Predicted Outputs)

A simple binary classification problem is used (spam = 1, not spam = 0) with 8 samples. The model output is a probability between 0 and 1 (sigmoid output). Three sets of predictions are used to see how each loss behaves.

Actual labels: [1, 0, 1, 1, 0, 0, 1, 0]

| Sample | Actual | Good | Poor | Confident but wrong |
|---|---|---|---|---|
| 1 | 1 | 0.90 | 0.40 | 0.90 |
| 2 | 0 | 0.10 | 0.60 | 0.10 |
| 3 | 1 | 0.80 | 0.30 | 0.80 |
| 4 | 1 | 0.95 | 0.55 | 0.95 |
| 5 | 0 | 0.20 | 0.70 | 0.20 |
| 6 | 0 | 0.05 | 0.45 | 0.05 |
| 7 | 1 | 0.85 | 0.35 | 0.85 |
| 8 | 0 | 0.15 | 0.65 | 0.99 |

## 4. Task 2: Formulas and Python Code

**Mean Squared Error:** MSE = (1/n) × Σ (y − ŷ)²

**Binary Cross-Entropy:** BCE = −(1/n) × Σ [ y·log(ŷ) + (1 − y)·log(1 − ŷ) ]

Here y is the actual label, ŷ is the predicted probability and n is the number of samples. Python code (NumPy):

In [1]:
"""
Deep Learning Assignment: Measuring prediction error with loss functions
Compares Mean Squared Error (MSE) and Binary Cross-Entropy (Log Loss)
on a simple binary classification problem (spam = 1, not spam = 0).
"""
import numpy as np

# ---------------------------------------------------------------
# Task 1: Actual class labels and predicted outputs
# ---------------------------------------------------------------
y_true = np.array([1, 0, 1, 1, 0, 0, 1, 0])

# Predicted probabilities from a model (sigmoid output, range 0-1)
y_pred_good = np.array([0.90, 0.10, 0.80, 0.95, 0.20, 0.05, 0.85, 0.15])  # confident and mostly right
y_pred_bad = np.array([0.40, 0.60, 0.30, 0.55, 0.70, 0.45, 0.35, 0.65])   # uncertain / mostly wrong
y_pred_confident_wrong = np.array([0.90, 0.10, 0.80, 0.95, 0.20, 0.05, 0.85, 0.99])  # last one: very wrong


# ---------------------------------------------------------------
# Task 2: Loss functions
# ---------------------------------------------------------------
def mean_squared_error(y_true, y_pred):
    """MSE = (1/n) * sum((y - y_hat)^2)"""
    return np.mean((y_true - y_pred) ** 2)


def binary_cross_entropy(y_true, y_pred, eps=1e-12):
    """BCE = -(1/n) * sum(y*log(y_hat) + (1-y)*log(1-y_hat))"""
    y_pred = np.clip(y_pred, eps, 1 - eps)  # avoid log(0)
    return -np.mean(y_true * np.log(y_pred) + (1 - y_true) * np.log(1 - y_pred))


# ---------------------------------------------------------------
# Task 3: Compare the results
# ---------------------------------------------------------------
cases = {
    "Good predictions": y_pred_good,
    "Poor predictions": y_pred_bad,
    "Confident but wrong (1 sample)": y_pred_confident_wrong,
}

print(f"{'Case':<34}{'MSE':>10}{'Cross-Entropy':>16}")
print("-" * 60)
for name, pred in cases.items():
    print(f"{name:<34}{mean_squared_error(y_true, pred):>10.4f}"
          f"{binary_cross_entropy(y_true, pred):>16.4f}")

# Per-sample view of one very wrong prediction (true label 0, predicted 0.99)
print("\nSingle sample: true = 0, predicted = 0.99")
print("  Squared error :", round((0 - 0.99) ** 2, 4))
print("  Cross-entropy :", round(-np.log(1 - 0.99), 4))

Case                                     MSE   Cross-Entropy
------------------------------------------------------------
Good predictions                      0.0187          0.1356
Poor predictions                      0.3688          0.9420
Confident but wrong (1 sample)        0.1384          0.6909

Single sample: true = 0, predicted = 0.99
  Squared error : 0.9801
  Cross-entropy : 4.6052


## 5. Output

The output of the program is shown below the code cell in Section 4:

```
Case                                     MSE   Cross-Entropy
------------------------------------------------------------
Good predictions                      0.0187          0.1356
Poor predictions                      0.3688          0.9420
Confident but wrong (1 sample)        0.1384          0.6909

Single sample: true = 0, predicted = 0.99
  Squared error : 0.9801
  Cross-entropy : 4.6052
```

## 6. Task 3: Comparison and Analysis

| Case | MSE | Cross-Entropy |
|---|---|---|
| Good predictions | 0.0187 | 0.1356 |
| Poor predictions | 0.3688 | 0.9420 |
| Confident but wrong (1 sample) | 0.1384 | 0.6909 |

**Observations:**

- Both loss functions give lower values when predictions are close to the actual labels and higher values for the poorer predictions in this experiment. They measure prediction error in different ways.
- Cross-Entropy penalizes confident incorrect predictions more strongly. For a sample with actual label 0 and predicted probability 0.99, the squared error is 0.9801, while the cross-entropy is approximately 4.6052. In this binary classification example, the squared error for one sample cannot exceed 1 because both the label and predicted probability are between 0 and 1. Cross-entropy, however, increases without bound as the model assigns a probability closer to 1 to the wrong class.
- With sigmoid or softmax outputs, cross-entropy generally provides more useful gradients for classification training than MSE, particularly when predictions are confidently incorrect. The exact gradient behavior depends on the model and loss implementation.
- Cross-Entropy is suited to predicted class probabilities and is commonly used in classification.

### When to use each loss function

| Loss function | Common use | Examples |
|---|---|---|
| Mean Squared Error (MSE) | Mainly used for regression, where the target is a continuous value; it can also be used in other settings | House-price or temperature prediction; autoencoder reconstruction |
| Cross-Entropy | Commonly used for classification with predicted probabilities | Binary classification with sigmoid output; multi-class classification with softmax output |

**Note:** Binary Cross-Entropy is used for binary classification, while Categorical Cross-Entropy is commonly used for multi-class classification.


## 7. Conclusion

Loss functions convert the difference between actual and predicted values into a number that a neural network aims to minimise during training. In this assignment, Mean Squared Error (MSE) and Binary Cross-Entropy were implemented in Python and compared using a simple binary classification problem. The results show that both functions measure prediction errors, but they respond differently to incorrect predictions. In particular, Binary Cross-Entropy penalizes highly confident incorrect probability predictions more strongly. MSE is mainly used for regression, while Cross-Entropy is commonly used for classification. Choosing a loss function depends on the type of task and model output.
